language/vm/src/file_format.rs

```diff
@@ -487,43 +486,123 @@ impl Signature {
 /// type parameter in the `FunctionHandle` and `StructHandle`.
 pub type TypeParameterIndex = u16;
 
-/// A `Kind` classifies types into sets with rules each set must follow.
+/// A `Constriant` classifies what operations are permitted for a given type
 ///
 /// Currently there are three kinds in Move: `All`, `Resource` and `Copyable`.
+#[repr(u8)]
 #[derive(Debug, Clone, Eq, Copy, Hash, Ord, PartialEq, PartialOrd)]
 #[cfg_attr(any(test, feature = "fuzzing"), derive(Arbitrary))]
-pub enum Kind {
-    /// Represents the super set of all types. The type might actually be a `Resource` or
-    /// `Copyable` A type might be in this set if it is not known to be a `Resource` or
-    /// `Copyable`
-    ///   - This occurs when there is a type parameter with this kind as a constraint
-    All,
-    /// `Resource` types must follow move semantics and various resource safety rules, namely:
-    /// - `Resource` values cannot be copied
-    /// - `Resource` values cannot be popped, i.e. they must be used
-    Resource,
-    /// `Copyable` types do not need to follow the `Resource` rules.
-    /// - `Copyable` values can be copied
-    /// - `Copyable` values can be popped
-    Copyable,
-}
-
-impl Kind {
-    /// Checks if the given kind is a sub-kind of another.
-    #[inline]
-    pub fn is_sub_kind_of(self, k: Kind) -> bool {
-        use Kind::*;
+pub enum Ability {
+    /// Allows values of that type to be copied, via CopyLoc or ReadRef
```

Comment:
Minor, but: does it make sense to say "of that type"? I think this is technically "types with this ability", but not sure I've mastered the terminology yet.

Reply:
I like  "types with this ability"